packages/tasit-action/src/TasitAct.test.js

```diff
@@ -0,0 +1,137 @@
+import { abi as contractABI } from "./helpers/SimpleStorage.json";
+
+// Note: Under the current `tasit-contracts` setup SimpleStorage aways will deployed with this address
+// See https://github.com/tasitlabs/TasitSDK/pull/59#discussion_r242258739
+const contractAddress = "0x6C4A015797DDDd87866451914eCe1e8b19261931";
+
+describe("Contract", function() {
+  let simpleStorage, wallet;
+
+  beforeEach("should connect to an existing contract", async () => {
+    // Account creates a wallet, should it create an account object that encapsulate the wallet?
+    // TasitAcount.create()
+    // > Acount { wallet: ..., metaTxInfos..., etc }
+    wallet = Account.createFromPrivateKey(
+      "0x11d943d7649fbdeb146dc57bd9cfc80b086bfab2330c7b25651dbaf382392f60"
+    );
+
+    simpleStorage = new Contract(contractAddress, contractABI);
+    expect(simpleStorage).to.exist;
+    expect(simpleStorage.address).to.equal(contractAddress);
+    expect(simpleStorage.getValue).to.exist;
+    expect(simpleStorage.setValue).to.exist;
+    // Events are not implemented yet
+    //expect(simpleStorage.ValueChanged).to.exist;
+  });
+
+  it("should throw error when instantiated with invalid args", async () => {
+    try {
+      new Contract();
+      assert(false, "constructor without address and ABI");
+
+      new Contract(contractAddress);
+      assert(false, "constructor without ABI");
+
+      new Contract("invalid address");
+      assert(false, "constructor with invalid address");
+
+      new Contract("invalid address", contractABI);
+      assert(false, "constructor with invalid address and valid ABI");
+
+      new Contract(contractAddress, "invalid abi");
+      assert(false, "constructor with valid address and invalid ABI");
+    } catch (e) {
+      assert(true);
+    }
+  });
+
+  it("should call a read-only contract method", async () => {
+    const value = await simpleStorage.getValue();
+    expect(value).to.exist;
+  });
+
+  it("should throw error when setting invalid wallet", async () => {
+    try {
+      simpleStorage = simpleStorage.setWallet();
+      assert(false, "setting no wallet");
+
+      simpleStorage = simpleStorage.setWallet("invalid wallet");
+      assert(false, "setting invalid wallet");
+    } catch (e) {
+      assert(true);
+    }
+  });
+
+  it("should throw error when calling write method without account/wallet/signer", async () => {
+    try {
+      simpleStorage.setValue("hello world");
+      assert(false, "calling write function without account");
+    } catch (e) {
+      assert(true);
+    }
+  });
+
+  it("should throw when subscribing with invalid trigger", async () => {
+    simpleStorage = simpleStorage.setWallet(wallet);
+    const subscription = simpleStorage.setValue("hello world");
+
+    try {
+      await subscription.on("invalid", () => {});
+      assert(false, "subscribing with invalid trigger");
+    } catch (e) {
+      await subscription.removeAllListeners();
```

Comment:
If we have proper teardown in-between tests, we shouldn't need to add this line directly here

Reply:
I think so. Did we also remove all the lines where we manually removed the listeners in a single test now that the `afterEach` covers it? You probably have done that for most or all of them already.

Reply:
Yes, I've kept `off()` function only inside test cases that the listener removal affects an expectation.

Reply:
Perfect, thanks. I'm resolving this.